keep/api/routes/rules.py

```diff
@@ -214,3 +228,206 @@ async def update_rule(
+                            }
+                        },
+                        "summery": {
+                            "type": "string",
+                            "description": "One liner summery of the results, mention what you noticed in the data and how you created the rules"
+                        }
+                    },
+                    "required": ["hasResults", "reason", "results", "summery"],
+                    "additionalProperties": False
+                }
+            }
+        ]
+
+
+@router.get(
+    "/gen_rules",
+    description="Generate Rules Using An AI",
+    status_code=202,
+)
+def gen_rules(
+    background_tasks: BackgroundTasks,
+    request: Request,
+    authenticated_entity: AuthenticatedEntity = Depends(
+        IdentityManagerFactory.get_auth_verifier(["read:rules", "read:alerts"])
+    ),
+):
+    
+    task_id = str(uuid.uuid4())
+    background_tasks.add_task(ruleGen, task_id, authenticated_entity)
+    return {"task_id": task_id}
+
+
+def ruleGen(task_id, authenticated_entity):
+
+    try:
+        logger.info(f"Generating rules for task {task_id}")
+
+        if "OPENAI_API_KEY" not in os.environ:
+            logger.error("OpenAI API key is not set. Can't auto gen rules.")
+            return ""
+
+        openAI_client = OpenAI(api_key=os.environ['OPENAI_API_KEY'])
+
+
+        existing_rules = get_rules(authenticated_entity)
+        existing_rules = [{'name': x['name'], 'cel_query': x['definition_cel'], 'group_by': x['grouping_criteria'], 'timeframe_mins' : x['timeframe']} for x in existing_rules]
+        existing_rules = 'here is a list of rules that already exist: ' + str(existing_rules)
+
+        tenant_id = authenticated_entity.tenant_id
+        
+        db_alerts = get_last_alerts(tenant_id=tenant_id, limit=ALERT_PULL_LIMIT)
+        db_alerts = [{'event' : x.event, 'timestamp' : x.timestamp.isoformat()} for x in db_alerts]
+        
+        selected_alerts = select_right_num_alerts(existing_rules, db_alerts, MAX_ALERT_TOKENS)
+        selected_alerts = "alert examples:" + json.dumps(selected_alerts)
+
+        response = openAI_client.chat.completions.create(
+            model = 'gpt-4o-mini',
+            messages = [{'role': 'system', 'content': SYSTEM_PROMPT}, 
+                        {'role': 'user', 'content': existing_rules},
+                        {'role': 'user', 'content': selected_alerts}],
+            functions = RESULT_CUSTOM_FUNCTION,
+            function_call = 'auto'
+        )
+        result = json.loads(response.choices[0].message.function_call.arguments)
+
+        logger.info("Got {} rules back from the llm".format(len(result['results'])))
+
+        result['results'] = [rule for rule in result['results'] if check_cel_rule(rule['CELRule'])]
+
+        
+        pusher_client = get_pusher_client()
+        if pusher_client:      
+            try:
+                pusher_client.trigger("private-{}".format(tenant_id), "rules-aigen-created", result)
```

Comment:
nit: `f"private-{tenant_id}"` is how I see things being done in the code.

Thinking out loud: does this trigger a websocket message to every user logged in with the same tenant?

Reply:
thats a great point, i stole pattern from some other parts of the codebase, should we do something about this? @shahargl